genie-core/src/main/java/com/netflix/genie/core/jpa/services/JpaClusterServiceImpl.java

```diff
@@ -398,69 +423,79 @@ public void addTagsForCluster(
     @Override
     @Transactional(readOnly = true)
     public Set<String> getTagsForCluster(
-        @NotBlank(message = "No cluster id sent. Cannot retrieve tags.")
-        final String id
+        @NotBlank(message = "No cluster id sent. Cannot retrieve tags.") final String id
     ) throws GenieException {
-        return this.findCluster(id).getTags();
+        return this.findCluster(id).getTags().stream().map(TagEntity::getTag).collect(Collectors.toSet());
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void updateTagsForCluster(
-        @NotBlank(message = "No cluster id entered. Unable to update tags.")
-        final String id,
-        @NotEmpty(message = "No tags entered. Unable to update.")
-        final Set<String> tags
+        @NotBlank(message = "No cluster id entered. Unable to update tags.") final String id,
+        @NotEmpty(message = "No tags entered. Unable to update.") final Set<String> tags
     ) throws GenieException {
-        this.findCluster(id).setTags(tags);
+        this.findCluster(id).setTags(this.createAndGetTagEntities(tags));
+        final ClusterEntity clusterEntity = this.findCluster(id);
+        final Set<TagEntity> newTags = this.createAndGetTagEntities(tags);
+        this.setFinalTags(newTags, clusterEntity.getUniqueId(), clusterEntity.getName());
+        clusterEntity.setTags(newTags);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void removeAllTagsForCluster(
-        @NotBlank(message = "No cluster id entered. Unable to remove tags.")
-        final String id
+        @NotBlank(message = "No cluster id entered. Unable to remove tags.") final String id
     ) throws GenieException {
-        this.findCluster(id).setTags(Sets.newHashSet());
+        final Set<TagEntity> tags = this.findCluster(id).getTags();
+        // Remove all the tags except the ones that start with "genie."
+        tags.removeAll(
+            tags
+                .stream()
+                .filter(
+                    tagEntity -> !tagEntity.getTag().startsWith(JpaBaseService.GENIE_TAG_NAMESPACE)
+                )
+                .collect(Collectors.toSet())
+        );
```

Comment:
Yes. Because the collection after exiting the method (and thus initiating a transaction commit) will only contain references to the ones it needs it will do a set differential and issue the proper queries.